src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt

```diff
@@ -146,14 +144,20 @@ class CodyToolWindowContent(private val project: Project) : UpdatableChat {
       promptPanel.textArea.emptyText.text = "Connecting to agent..."
     }
 
-    ApplicationManager.getApplication().executeOnPooledThread {
-      getInitializedServer(project).thenAccept { server ->
-        id = server.chatNew().get()
+    CodyAgentService.applyAgentOnBackgroundThread(project) { agent ->
+      try {
+        id = agent.server.chatNew().get()
         ApplicationManager.getApplication().invokeLater {
           promptPanel.textArea.isEnabled = true
           promptPanel.textArea.emptyText.text = "Ask a question about this code..."
         }
         callback.invoke()
+      } catch (e: ExecutionException) {
+        // Agent cannot gracefully recover when connection is lost, we need to restart it
+        logger.warn("Failed to load new chat, restarting agent", e)
+        CodyAgentService.getInstance(project).restartAgent(project)
+        Thread.sleep(5000)
```

Comment:
is it needed to sleep here? `startAgent` has it's 15s timeout already ⌚

Reply:
I wanted to avoid constant heavier utilisation of the CPU in case of the lack of the network.
In fact it would be better to have some listener which would say if network is back, I will check if we can do that. I will also move that code to actor service.